hotp/ta/hotp_ta.c

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <hotp_ta.h>
+#include <string.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+
+/*
+ * This macro checks whether the return value is not equal to TEE_SUCCESS, if it
+ * is not equal, then it will set the return value to 'res' and then goto
+ * 'label'.
+ */
+#define CHECK_EXIT(label, res, ret_val) \
+	if (ret_val != TEE_SUCCESS) { \
+		EMSG("0x%08x", ret_val); \
+		res = ret_val; \
+		goto label; }
+
+/*
+ * This macro checks the expected parameters, if they are not the same, then it
+ * will set 'res' to TEE_ERROR_BAD_PARAMETERS and will goto label 'label'.
+ */
+#define CHECK_EXP_PARAM_EXIT(label, res, pt, ept) \
+	if (pt != ept) { \
+		EMSG("Expected: 0x%x, got: 0x%x", ept, pt); \
+		res = TEE_ERROR_BAD_PARAMETERS; \
+		goto label; }
+
+#define SWAP64(v) \
```

Comment:
We already have such a macro in a header accessible to TAs, if I'm not mistaken.

Reply:
I think I tried using `TEE_U64_BSWAP`, but that didn't work I think (only visible to utee?). Anyhow, it's not needed / used anymore since I went back to my old way of doing the increment.